feat(minifier): implement dot define for member expressions

camc314 opened the pull request:
adds the ability to transform

`process.env.NODE_ENV`

into 

`"development"`

when that config is passed in

camc314 commented:
* **#3959** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3959?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camc314 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

camc314 commented on crates/oxc_minifier/src/ast_passes/replace_global_defines.rs:
```diff
@@ -84,8 +113,8 @@ impl<'a> ReplaceGlobalDefines<'a> {
     fn replace_identifier_defines(&self, expr: &mut Expression<'a>) {
-        for (key, value) in &self.config.0.identifier_defines {
-            if let Expression::Identifier(ident) = expr {
+        if let Expression::Identifier(ident) = expr {
+            for (key, value) in &self.config.0.identifier_defines {
```
i changed this to avoid looping over `identifier_defines` if we are not processing an identifier

camc314 commented on crates/oxc_minifier/src/ast_passes/replace_global_defines.rs:
```diff
@@ -19,7 +19,12 @@ pub struct ReplaceGlobalDefinesConfig(Arc<ReplaceGlobalDefinesConfigImpl>);
 #[derive(Debug)]
 struct ReplaceGlobalDefinesConfigImpl {
     identifier_defines: Vec<(/* key */ String, /* value */ String)>,
-    // TODO: dot defines
+    dot_defines: Vec<(/* member expression parts */ Vec<String>, /* value */ String)>,
```
these can probably be optimized into hashmaps.

changing them into hash maps will also allow us to detect collisions easier

e.g. `ReplaceGlobalDefinesConfig::new(&[("foo", "bar"), ("foo","baz")])`

should probably report an error since `foo=baz` will never be respected

DonIsaac reviewed (commented):
Could you please add some test cases for this PR?

camc314 commented:
> Could you please add some test cases for this PR?

there should be some in here?

https://github.com/oxc-project/oxc/blob/1e77217321a64508774bac76faa904157aa6c6e8/crates/oxc_minifier/tests/oxc/replace_global_defines.rs#L26-L40

or do you mean more integration-esque? if so where do those live?

DonIsaac commented:
> > Could you please add some test cases for this PR?
> 
> 
> 
> there should be some in here?
> 
> 
> 
> https://github.com/oxc-project/oxc/blob/1e77217321a64508774bac76faa904157aa6c6e8/crates/oxc_minifier/tests/oxc/replace_global_defines.rs#L26-L40
> 
> 
> 
> or do you mean more integration-esque? if so where do those live?

Forgive me, I didn't see those! That's perfect, thanks.

Boshen commented:
I'll revisit this PR with the one in Rolldown together this week.

camc314 commented:
> I'll revisit this PR with the one in Rolldown together this week.

sounds good. feel free to close if you don't think this is the right approach

camc314 committed:
feat(minifier): implement dot define for member expressions (#3959)

adds the ability to transform

`process.env.NODE_ENV`

into

`"development"`

when that config is passed in
